kpet/data.py

```diff
@@ -451,6 +456,70 @@ def convert_tree_arches(self):
 
             self.trees[name]["arches"] = list(tree_arches)
 
+    def convert_sets(self):
+        """
+        Convert each suite / cases sets from a list of regexes
+        to a list of set names matching those regexes
+
+        Raises:
+            A schema.Invalid exception when finding an invalid regex
+        """
+
+        error = ("One of '{0}: set:' regexes\n" +
+                 "doesn't match any of the available sets\n" +
+                 "The regex: '{1}'\n" +
+                 "The avaliable sets: " + str(list(self.sets.keys())))
+
+        # TODO: Remove function after conversion
+        # This function is to convert from the old format (where sets were a
+        # part of the pattern)
+        def convert_pattern(pattern):
+            sets = set(self.sets.keys())
+            matches = set()
+            for set_name in sets:
+                target = Target(trees=Target.ANY,
+                                arches=Target.ANY,
+                                components=Target.ANY,
+                                sets=set([set_name]),
+                                sources=Target.ANY)
+                if pattern.matches(target):
+                    matches.add(set_name)
+            if matches == sets:
+                # If a pattern matches every set it should mean it doesn't have
+                # them specified
+                return set()
+            return matches
+
+        for suite in self.suites:
+            suite_sets = set()
+            for regex in suite.sets or set():
+                match = set(filter(regex.fullmatch, self.sets.keys()))
+                if match == set():
+                    raise Invalid(error.format("suite", regex.pattern))
+                suite_sets |= match
+
+            suite.sets = suite_sets | convert_pattern(suite.pattern)
+
+            for case in suite.cases:
+                case_sets = set()
+                for regex in case.sets or set():
+                    match = set(filter(regex.fullmatch, self.sets.keys()))
+                    if match == set():
+                        raise Invalid(error.format("case", regex.pattern))
+                    case_sets |= match
+
+                case.sets = case_sets | convert_pattern(case.pattern)
+
+                if case.sets == set():
+                    # If the case doesn't specify any of it's own sets,
+                    # we inherit the suite's sets
+                    case.sets = suite.sets
```

Comment:
I thought we don't want the former to happen because that'd mean we have an 'unassigned' case.

Either way, with the way sets are done now I'm not sure you can specify a "empty list" in them. Perhaps I missed that this should be possilbe? Or maybe I'm doing it wrong but just
`sets:` won't work

Reply:
mb, haven't checked the new commits yet ^